### DISCUSSION TITLE
@percy/cli not converting canvas to img => nothing displaying

### DISCUSSION THREAD

**Discussion by emilgoldsmith** [percy/cli#715] [Support] [answered]
## Have you read the "Debugging SDKs" doc yet?

Yes

## The problem

With @percy/agent my WebGL canvases are now successfully working with the help from your debugging SDKs resource, but when I upgraded to @percy/cli it broke sadly. When I do the "download source" option I can see that the main difference is that with @percy/agent the canvases have been converted to images as you also describe in the debugging SDKs document that is the intention. With @percy/cli however they are still canvases... empty canvases sadly. It could of course be you changed methods so you no longer need to convert to an  image but in that case that seems broken.

By the way I am using the cli command `percy snapshot` directly with the `@percy/cli` package, no intermediaries

## Environment

- Node version:
- `@percy/cli` version: 1.0.0-beta.71
- Version of Percy SDK you’re using: N/A
- If needed, a build or snapshot ID:
- OS version: Ubuntu 20.04
- Type of shell command-line [interface]: Bash

## Details

If necessary, describe the problem you have been experiencing in more detail.

## Debug logs

```
[percy:config] Config file not found (0ms)
[percy:core:install] Successfully downloaded Chromium 885264 (6875ms)
[percy:core:browser] Launching browser (21ms)
[percy:core:browser] Browser connected: HeadlessChrome/92.0.4515.0 (172ms)
[percy:core] Percy has started! (0ms)
[percy:core:snapshot] --------- (4ms)
[percy:core:snapshot] Handling snapshot: /AlgorithmFromString.elm.html (1ms)
[percy:core:snapshot] - url: http://localhost:43785/AlgorithmFromString.elm.html (0ms)
[percy:core:snapshot] - widths: 375px, 1280px (0ms)
[percy:core:snapshot] - minHeight: 1024px (0ms)
[percy:core:snapshot] - discovery.allowedHostnames: localhost (0ms)
[percy:core:snapshot] --------- (1ms)
[percy:core:snapshot] Handling snapshot: /CubeViews.elm.html (0ms)
[percy:core:snapshot] - url: http://localhost:43785/CubeViews.elm.html (0ms)
[percy:core:snapshot] - widths: 375px, 1280px (0ms)
[percy:core:snapshot] - minHeight: 1024px (0ms)
[percy:core:snapshot] - discovery.allowedHostnames: localhost (0ms)
[percy:core:page] Page created (7ms)
[percy:core:page] Page created (4ms)
[percy:core:page] Resize page to 375x1024 (21ms)
[percy:core:page] Resize page to 375x1024 (0ms)
[percy:core:page] Navigate to: http://localhost:43785/CubeViews.elm.html (2ms)
[percy:core:page] Navigate to: http://localhost:43785/AlgorithmFromString.elm.html (1ms)
[percy:core:discovery] Handling request: http://localhost:43785/CubeViews.elm.html (7ms)
[percy:core:discovery] Handling request: http://localhost:43785/AlgorithmFromString.elm.html (0ms)
[percy:core:page] Page navigated (81ms)
[percy:core:network] Wait for 100ms idle (1ms)
[percy:core:discovery] Processing resource: http://localhost:43785/AlgorithmFromString.elm.html (7ms)
[percy:core:discovery] - sha: 27c0813a33bb6010c44e9050b16ade7dd1dfdaf2da1ef68eca7b343fb539f82d (1ms)
[percy:core:discovery] - mimetype: text/html (0ms)
[percy:core:page] Resize page to 1280x1024 (103ms)
[percy:core:page] Taking snapshot: /AlgorithmFromString.elm.html (4ms)
[percy:core:network] Wait for 100ms idle (0ms)
[percy:core:page] Page navigated (91ms)
[percy:core:network] Wait for 100ms idle (0ms)
[percy:core:page] Inject @percy/dom (12ms)
[percy:core:page] Serialize DOM (6ms)
[percy:core] Snapshot taken: /AlgorithmFromString.elm.html (14ms)
[percy:core:page] Page closed (4ms)
[percy:core:discovery] Processing resource: http://localhost:43785/CubeViews.elm.html (583ms)
[percy:core:discovery] - sha: 2eecae26961b7f8e3e5f178a47a1a226ef52a907737bda80d0b67cb0c79cce4c (0ms)
[percy:core:discovery] - mimetype: text/html (0ms)
[percy:core:page] Resize page to 1280x1024 (101ms)
[percy:core:page] Taking snapshot: /CubeViews.elm.html (3ms)
[percy:core:network] Wait for 100ms idle (0ms)
[percy:core:page] Inject @percy/dom (102ms)
[percy:core:page] Serialize DOM (4ms)
[percy:core] Snapshot taken: /CubeViews.elm.html (31ms)
[percy:core:page] Page closed (2ms)
[percy:core:browser] Closing browser (15ms)
[percy:core] Build not created (34ms)
```

## Code to reproduce issue

This is the html document I am snapshotting using `percy snapshot` https://pastebin.com/TQBZ0F96

**Comment by Robdel12** [comment]
Hey @emilgoldsmith! It's the same method (https://github.com/percy/cli/blob/master/packages/dom/src/serialize-canvas.js). Looking at a build, it looks like it's capturing an empty image: 

<img width="548" alt="image" src="https://user-images.githubusercontent.com/2072894/148814880-fa00d05b-de49-47e5-8cf8-965a4d482b31.png">

Which suggests the drawingBuffer is still empty [when `canvas.toDataUrl()` is called](https://github.com/percy/cli/blob/master/packages/dom/src/serialize-canvas.js#L7)

**Comment by emilgoldsmith** [comment]
Hmmm... Well how does that explain that it works with `@percy/agent` though? There must be something going wrong somewhere. If not an actual bug in the CLI then at least some different configuration that is needed in the new version of the CLI

**Reply by Robdel12** [reply]
Implementation didn't change but what likely did change is timing in your suite when the snapshot is captured. You might need to use a wait to make sure the page/canvas element is in the right state before a snapshot is captured: https://docs.percy.io/docs/percy-snapshot#advanced 

If you need to run JS prior to the snapshot being captured to prepare the canvas element, that can also be done: https://docs.percy.io/docs/percy-snapshot#executing-javascript-after-events

**Comment by emilgoldsmith** [comment]
That sounds very promising! Thanks Robert, I’ll give that a try :)

**Comment by emilgoldsmith** [comment]
Sadly that still didn't work @Robdel12 :(. Since I couldn't specify waitForTimeout with the `directory` argument I rewrote that part of my CI to start a static server and use the `snapshots.yml` method described in the docs with a `waitForTimeout` option and even had it set to 10 seconds, and while it does now convert it to an image it seems the image is empty, while `@percy/agent` still works for me, so not quite sure what else I can try

**Reply by Robdel12** [reply]
It might not need _more_ time. Since it worked with agent, the thing that changed was timing (CLI will likely wait longer prior to snapshot, it's controlled by the asset discovery timeout now). 

Whatever it might be, the canvas elements drawing buffer is wiped by the time the snapshot is captured. I'm pretty sure if you checked the canvas element prior to the snapshot being captured, `canvas.toDataURL()` would also return a base64 string of an empty png. https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL

**Reply by emilgoldsmith** [reply]
@Robdel12, thanks for sticking with me and keeping up suggestions Robert!

I of course can't say for sure that I can rule that out but I think it's pretty likely as I just tried opening the html file in my browser and then doing `toDataURL` and I got something non-empty
![Screenshot from 2022-01-15 19-04-23](https://user-images.githubusercontent.com/18149242/149632832-94a3af37-0b85-4c28-b664-5c9939809554.png)

I would say this is pretty decent proof that given enough time that should also work? Of course I can't say for sure, but I feel like it's pretty strong evidence.

For now I'll just keep using `@percy/agent` and that works well for me. Otherwise I know it can also work with Cypress but that just seems like a lot of overhead for a simple task that I don't need in this library. If you have other suggestions I'm happy to try them though

**Reply by Robdel12** [reply]
No problem! 

> and I got something non-empty

That doesn't mean it's not an empty transparent png though. If you put that into an img tag, what renders? If you look at the first snapshot I debugged with blank canvas elements, there's a full base64 encoded string in the img tag, but it's a blank transparent PNG (because that's what is present in the drawing buffer at the time `toDataURL` is called)

**Comment by emilgoldsmith** [comment, marked as answer]
Sorry I should’ve mentioned more specifically that it was the png not just
the string that was empty, checked it properly, it displays the intended
image

**Reply by Robdel12** [reply]
After some digging, it looks like with this page it only will capture the canvas in an image if the browser is headed. This is what it took for me to get it to work:

```yml
# .percy.yml
version: 2
discovery:
  launch-options:
    headless: false
```

**Comment by emilgoldsmith** [comment]
How peculiar, but thank you so much for digging into it for me Robert! I
hope you have a good day!
